### DISCUSSION TITLE
Hint a manager that a different type of versioning engine has to be used.

### DISCUSSION THREAD

**Discussion by BobVanB** [renovatebot/renovate#18498] [Request Help]
## Request

I do not have a `-SNAPSHOT` in my version and renovate proposes `-SNAPSHOT`.
I would like to have only the 7.17.3 updated to 7.17.7.

`lib/module/versioning/docker/index.spec.ts`

```jest
    ✕ isStable("3.8.3-SNAPSHOT") === false (2 ms)
    ✕ isStable("3.8.3-1-SNAPSHOT") === false (1 ms)
```

My suggestion is to hint to the renovate engine that i want to use the versioning of `lib/module/versioning/maven`<br/>
Same test as above:

```jest
PASS  lib/modules/versioning/maven/index.spec.ts
```

## Demo

Pull request: https://github.com/renovatebot/renovate/pull/18472

## Context

Renovate version:<br/>
`renovate/renovate:32.165.1-slim@sha256:fbe8b824c6d123fe7f28bed65fecdd7b8bb57e6a1eb513afcf8bc862e43b8993`

```json
               {
                 "depName": "docker.elastic.co/kibana/kibana",
                 "currentValue": "7.17.3",
                 "datasource": "docker",
                 "versioning": "regex:^[^\\d]*(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*-SNAPSHOT)|-(?<build>.*))?.*$",
                 "replaceString": "\n    image: docker.elastic.co/kibana/kibana:7.17.3",
                 "depIndex": 2,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "7.17.7-SNAPSHOT",
                     "newValue": "7.17.7-SNAPSHOT",
                     "newMajor": 7,
                     "newMinor": 17,
                     "updateType": "patch",
                     "branchName": "renovate/elastic-lifecycle"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "8.6.0-SNAPSHOT",
                     "newValue": "8.6.0-SNAPSHOT",
                     "newMajor": 8,
                     "newMinor": 6,
                     "updateType": "major",
                     "branchName": "renovate/major-elastic-lifecycle"
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/elastic/kibana",
                 "currentVersion": "7.17.3",
                 "isSingleVersion": true,
                 "fixedVersion": "7.17.3"
               }
```

### If you have any ideas on how this should be implemented, please tell us here.

Sorry i don't know.

### Is this a feature you are interested in implementing yourself?

No

**Comment by viceice** [comment]
You've explict set the versioning to regex, which allowes the snapshot versions. you should use the default docker versioning for docker images.

Please provide your config, so we can optimize it.

**Reply by BobVanB** [reply]
The default versioningTemplate is used to support our own template:

* image:name-v1.2.3
* image:name-v1.2.3-4
* value: v1.3.3
* version: v1.3.3
* ...

Renovate Regex Manager
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["..."],
  "regexManagers": [
    {
      "fileMatch": ["(^|/)kustomization\\.ya?ml$"],
      "matchStrings": [
        "# renovate: depName=(?<depName>[^\\s]+?)(?: (lookupName|packageName)=(?<packageName>[^\\s]+?))?(?: versioning=(?<versioning>[a-z-0-9]+?))?\\s*(value|version):\\s\"?(?<currentValue>.+?)\"?\\s",
        "\\s*image:\\s*\\\"?(?<depName>.*):(?<currentValue>.+)\\\"?"
      ],
      "versioningTemplate": "regex:^[^\\d]*(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*-SNAPSHOT)|-(?<build>.*))?.*$",
      "datasourceTemplate": "docker"
    },
```

Renovate config:
```json
{
  "branchConcurrentLimit": 0,
  "prHourlyLimit": 0,
  "prConcurrentLimit": 0,
  "assignees": ["..."],
  "assigneesSampleSize": 1,
  "autodiscover": true,
  "autodiscoverFilter": "...",
  "includeForks": true,
  "ignoreUnstable": true,
  "onboarding": true,
  "onboardingConfig": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [ "local>renovate/renovate-runner:base" ]
  }
}
```

Example kustomization.yaml that contains one of each matchStrings:
```
patchesStrategicMerge:
- |-
  apiVersion: ...
  kind: ...
  metadata:
    name: ...
  spec:
    # renovate: depName=kibana
    version: 7.17.3
    image: docker.elastic.co/kibana/kibana:7.17.3
```

**Reply by BobVanB** [reply]
Just to clarify.

The first match of the same regex, is updated to 7.17.6.
```yaml
    # renovate: depName=kibana
    version: 7.17.3
```
Only the second part is going wrong
```yaml
    image: docker.elastic.co/kibana/kibana:7.17.3
```

> You've explict set the versioning to regex, which allowes the snapshot versions. you should use the default docker versioning for docker images.

* Is setting the ` datasourceTemplate` only telling renovate to use the docker versioning for the source or is it also used for the remote matches?
* Why are the tags from the `sourceUrl` used and not the tags from `docker.elastic.co`
  Because there are no `-SNAPSHOT` tags in there.
* If docker is used and the tags come from a git repository, renovate should switch versioning to something else.
  In this case, maven versioning. That is the only thing that supports `-SNAPSHOTS`.
  * Renovate should detect the type of versioning used, if it is switching to a git repository. (should do this anyway)
  * I want to set a parameters and telling renovate to use a different versioning enging. In this case maven.
* If possible renovate should detect the parameters from the source and match this in the remote versions.
  Meening: If a match group in the regex is nill/null/empty, just don't update `release` or `build`
* The default docker tags are not that elaborate, for maven builds we can have images in our private repository with tags that end up with `-SNAPSHOT`

**Reply by viceice** [reply]
you should remove versioning form regex manager and configure it via package rule to only the packages which need that and use default versioning for the others.

the kustomize manager is probably interfere with your regex manager.
